x/proof/keeper/msg_server_submit_proof.go

```diff
@@ -127,12 +224,141 @@ func (k msgServer) queryAndValidateClaimForProof(ctx context.Context, proof *typ
 
 	// Ensure service IDs match.
 	if claimSessionHeader.GetService().GetId() != proofSessionHeader.GetService().GetId() {
-		return types.ErrProofInvalidService.Wrapf(
+		return nil, types.ErrProofInvalidService.Wrapf(
 			"claim service ID %q does not match proof service ID %q",
 			claimSessionHeader.GetService().GetId(),
 			proofSessionHeader.GetService().GetId(),
 		)
 	}
 
+	return &foundClaim, nil
+}
+
+// compareSessionHeaders compares a session header against an expected session header.
+func compareSessionHeaders(
+	expectedSessionHeader *sessiontypes.SessionHeader,
+	sessionHeader *sessiontypes.SessionHeader,
+) error {
+	if sessionHeader.GetApplicationAddress() != expectedSessionHeader.GetApplicationAddress() {
+		return types.ErrProofInvalidRelay.Wrapf(
+			"sessionHeaders application addresses mismatch expect: %q, got: %q",
+			expectedSessionHeader.GetApplicationAddress(),
+			sessionHeader.GetApplicationAddress(),
+		)
+	}
+
+	if sessionHeader.GetService().GetId() != expectedSessionHeader.GetService().GetId() {
```

Comment:
No strong opinion here, but since one side of the comparison is an off-chain header, we could ensure that it was properly constructed by comparing the service name too.

Will add it